docs/madr/decisions/046-meshservice-policy-matching.md

```diff
@@ -0,0 +1,174 @@
+semantic meaning, beyond other arbitrary, user-defined labels.
+
+## Decision Outcome
+
+These new kinds allow for selecting or directing traffic that's going to new,
+`MeshService`-based destinations.
+
+### `targetRef`
+
+The policy is applied
+on both the Envoy outbound listeners and clusters
+that correspond to the `MeshService`.
+
+References for `kind: MeshService` have the following structure:
+
+```yaml
+kind: MeshService
+name: backend
+labels: {}
+zone: east
+namespace: backend
+port: 80
+```
+
+In general, we try to do the expected thing and refer to `MeshService` via its display name.
+
+Exactly one of either `name` or `labels` is **required**.
+
+#### `port`
+
+If `port` is set in `targetRef.kind`, only traffic to that port is affected.
+
+In `backendRefs`, `port` is required.
+
+#### Local `MeshServices`
+
+If the policy ref doesn't set `zone`:
+
+* without `namespace` set on a k8s zone:
+  * if the policy is *not in `kuma-system`*, the ref refers to
+    `MeshServices` in the same namespace
+  * if the policy is in `kuma-system`, the ref refers to
+      the core name of the `MeshService`
+* with `namespace`, only that namespace is matched for `MeshServices`
+
+#### Non-local `MeshServices`
+
+If `zone` is set, `name` refers to the name of the `MeshService` in that zone.
+
+Without `namespace`, refer to _all_ `MeshServices` with this name. That means we
+don't infer any similarity between the local namespace and the other zone's
+namespace.
+
+#### Examples
+
+##### `MeshService`
+
+Users can target `MeshServices` directly by `name`:
+
+```yaml
+spec:
+ to:
+ - targetRef:
+     kind: MeshService
+     name: backend
+```
+
+or by using `labels` without `name`:
+
+```yaml
+spec:
+ to:
+ - targetRef:
+     kind: MeshService
+     labels:
+       kubernetes.io/service-name: zk
+```
+
+##### `backendRef`
+
+The structure of `backendRef` for `MeshService` is largely similar except that
+we require a specific port.
+
+```yaml
+kind: MeshHTTPRoute
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+   - targetRef:
+       kind: Mesh
+     rules:
+       - matches:
+           - path:
+               type: Prefix
+               value: /v1
+         default:
+           backendRefs:
+             - kind: MeshService
+               name: frontend
+               port: 8080
+               zone: east
+             - kind: MeshMultiZoneService
+               name: frontend
+               port: 8080
+```
+
+#### Status
+
+Similar to Gateway API conditions, we report back as status conditions
+if a route targets a
+`MeshService`/port tuple that doesn't exist.
+
+### Positive Consequences
+
+* Allows users to apply policy to traffic going to `MeshService` and
+  `MeshMultiZoneService`
+* We can match set of `MeshServices`, important for headless services
+* We're flexible with regards to local vs non-local `MeshServices`.
+
+### Negative Consequences
+
+* Relative complexity of the semantics of various combinations
```

Comment:
how do we select multiple services? Can MeshService just select multiple services or do we introduce MeshServiceGroup?

Reply:
It should be covered by using `labels` instead of `name`

Reply:
It's also not super clear where you can use a set of services vs a single service

Reply:
Do we need to differentiate it?

Reply:
To be clear, I don't mind if we want to introduce another kind for groups, as long as the naming is consistent.

Reply:
I'm concerned with the same targetRef and merging. How do you order policies?

Reply:
If you order MeshServiceGroup before MeshService, you order MeshService + labels before MeshService + name.
There's no information lost between the two representations.